File: scala/1975.maximum-matrix-sum.scala
/*
 * @lc app=leetcode id=1975 lang=scala
 *
 * [1975] Maximum Matrix Sum
 */

// @lc code=start
object Solution {
  def maxMatrixSum(matrix: Array[Array[Int]]): Long = {
    var totalAbsoluteSum: Long = 0
    var minAbs: Int = Int.MaxValue
    var negativeCount: Int = 0

    matrix.flatten.foreach { num =>
      totalAbsoluteSum += math.abs(num)
      if (num < 0) negativeCount += 1
      minAbs = math.min(minAbs, math.abs(num))
    }

    if (negativeCount % 2 == 0) totalAbsoluteSum
    else totalAbsoluteSum - 2 * minAbs
  }
}

// @lc code=end

File: python/1975.maximum-matrix-sum.py
#
# @lc app=leetcode id=1975 lang=python3
#
# [1975] Maximum Matrix Sum
#

# @lc code=start
class Solution:
    def maxMatrixSum(self, matrix: List[List[int]]) -> int:
        totalAbsoluteSum = 0
        totalNegativeCount = 0
        minPositive = float('inf')

        for row in matrix:
            for num in row:
                totalAbsoluteSum += abs(num)
                if num < 0:
                    totalNegativeCount += 1
                    minPositive = min(minPositive, abs(num))
                else:
                    minPositive = min(minPositive, num)

        if totalNegativeCount % 2 == 0:
            return totalAbsoluteSum
        else:
            return totalAbsoluteSum - 2 * minPositive
# @lc code=end

